### PR TITLE
Fix creating OAuth2 auth source from CLI

### PR DESCRIPTION
**Pull request by daniil-pankratov**
This MR fixing creation OAuth2 auth source from CLI. 
Problem is described here: https://github.com/go-gitea/gitea/issues/8356

### CODE REVIEW

File: routers/user/auth.go [hunk 1]
```diff
@@ -570,8 +570,15 @@ func SignInOAuth(ctx *context.Context) {
 		return
 	}
 
-	err = oauth2.Auth(loginSource.Name, ctx.Req.Request, ctx.Resp)
-	if err != nil {
+	if err = oauth2.Auth(loginSource.Name, ctx.Req.Request, ctx.Resp); err != nil {
+		if strings.Contains(err.Error(), "no provider for ") {
+			if err = models.ResetOAuth2(); err != nil {
+				ctx.ServerError("SignIn", err)
```
Comment (lunny) [inline routers/user/auth.go:576]: return

File: routers/user/auth.go [hunk 2]
```diff
@@ -570,8 +570,15 @@ func SignInOAuth(ctx *context.Context) {
...
+			}
+			if err = oauth2.Auth(loginSource.Name, ctx.Req.Request, ctx.Resp); err != nil {
+				ctx.ServerError("SignIn", err)
```
Comment (lunny) [inline routers/user/auth.go:580]: As above

File: routers/user/auth.go (hunk 1 above)
Comment (daniil-pankratov) [reply routers/user/auth.go:576]: Fixed

File: routers/user/auth.go (hunk 2 above)
Comment (daniil-pankratov) [reply routers/user/auth.go:580]: Fixed

File: models/oauth2.go [hunk 3]
```diff
@@ -125,8 +125,18 @@ func InitOAuth2() error {
...
+}
 
+// ResetOAuth2 clears existing OAuth2 providers and loads them from the DB
+func ResetOAuth2() error {
+	oauth2.ClearProviders()
+	return initOuath2LoginSources()
+}
+
+// initOuath2LoginSources is used to load and register all active OAuth2 providers
+func initOuath2LoginSources() error {
```
Comment (noerw) [inline models/oauth2.go]: typo, should be `initOAuth2LoginSources`

Comment (daniil-pankratov) [reply models/oauth2.go]: Fixed

**Review by noerw — APPROVED** [review]
Thanks for the PR! (fixes #8356 and #12938)

It would be nicer if `goth` providers were updated when *adding* a provider via CLI, instead of an error fallback during login.
The root of the issue is, that the CLI modifies the state of `goth` in the CLI process, not in the long-running gitea process.
Currently there is no RPC mechanism for this, we could add an API endpoint `POST /api/v1/admin/auth-source` for this.

As a first stopgap for the bug this still looks good to me.

**Review by lunny — APPROVED** [review]

**Comment by zeripath** [comment]
make lg-tm work

**Pull request merged**